File: examples/echo-agent/echo_agent.py
"""Minimal LAP Local 0.1 reference agent; stdout contains protocol frames only."""
from __future__ import annotations

import json
import sys
from typing import Any


sequence = 0


def emit(message_type: str, payload: dict[str, Any], *, correlation_id: str | None = None,
         run: dict[str, Any] | None = None) -> None:
    global sequence
    sequence += 1
    frame: dict[str, Any] = {
        "lap": "0.1",
        "id": f"echo-{sequence}",
        "producer": "org.lap.echo-agent",
        "seq": sequence,
        "type": message_type,
        "payload": payload,
    }
    if correlation_id:
        frame["correlation_id"] = correlation_id
    if run:
        frame["run"] = run
    print(json.dumps(frame, ensure_ascii=False, separators=(",", ":")), flush=True)


def handle(frame: dict[str, Any]) -> bool:
    message_type = frame.get("type")
    if message_type == "agent.hello":
        emit("agent.welcome", {
            "selected_lap": "0.1",
            "profiles": ["lap-local/0.1"],
            "agent_id": "org.lap.echo-agent",
            "version": "0.1.0",
            "max_concurrency": 1,
        }, correlation_id=str(frame.get("id") or ""))
        return True
    if message_type == "run.start":
        run = frame.get("run") or {}
        emit("run.accepted", {"capability": frame.get("payload", {}).get("capability")},
             correlation_id=str(frame.get("id") or ""), run=run)
        emit("run.progress", {"phase": "agent", "message": "Echoing input."}, run=run)
        text = str((frame.get("payload", {}).get("input") or {}).get("text") or "")
        emit("run.result", {"status": "succeeded", "summary": "Echo complete.",
             "output": {"text": text}}, run=run)
        return True
    if message_type == "run.cancel":
        emit("run.result", {"status": "cancelled", "summary": "Run cancelled.",
             "error": {"code": "LAP-401", "message": "Cancellation requested.",
                       "retryable": False}}, run=frame.get("run") or {})
        return True
    if message_type == "agent.shutdown":
        return False
    return True


for raw in sys.stdin:
    try:
        message = json.loads(raw)
        if not isinstance(message, dict):
            raise ValueError("frame must be an object")
    except (json.JSONDecodeError, ValueError):
        # A production Host terminates malformed peers; this example exits cleanly.
        sys.exit(2)
    if not handle(message):
        break

File: examples/echo-agent-node/echo_agent.js
// LAP Node.js Echo Agent. Stdout is protocol-only.
"use strict";

const readline = require("node:readline");

const AGENT_ID = "org.lap.echo-agent-node";
const AGENT_VERSION = "0.1.0";
let sequence = 0;

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function asObject(value) {
  return isObject(value) ? value : {};
}

function emit(messageType, payload, correlationId = "", run = null) {
  sequence += 1;
  const frame = {
    lap: "0.1",
    id: `node-echo-${sequence}`,
    producer: AGENT_ID,
    seq: sequence,
    type: messageType,
    payload,
  };
  if (correlationId) frame.correlation_id = correlationId;
  if (isObject(run)) frame.run = run;
  process.stdout.write(`${JSON.stringify(frame)}\n`);
}

function handle(frame) {
  const messageType = frame.type;
  if (messageType === "agent.hello") {
    emit("agent.welcome", {
      selected_lap: "0.1",
      profiles: ["lap-local/0.1"],
      agent_id: AGENT_ID,
      version: AGENT_VERSION,
      max_concurrency: 1,
    }, String(frame.id || ""));
    return true;
  }
  if (messageType === "run.start") {
    const run = asObject(frame.run);
    const payload = asObject(frame.payload);
    const input = asObject(payload.input);
    const text = typeof input.text === "string" ? input.text : "";
    emit("run.accepted", { capability: payload.capability }, String(frame.id || ""), run);
    emit("run.progress", { phase: "agent", message: "Processing requested task." }, "", run);
    emit("run.result", {
      status: "succeeded",
      summary: "Echo complete.",
      output: { text },
    }, "", run);
    return true;
  }
  if (messageType === "run.cancel") {
    emit("run.result", {
      status: "cancelled",
      summary: "Run cancelled.",
      error: { code: "LAP-401", message: "Cancellation requested.", retryable: false },
    }, "", asObject(frame.run));
    return true;
  }
  return messageType !== "agent.shutdown";
}

const input = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
let stopped = false;
input.on("line", (raw) => {
  if (stopped) return;
  let frame;
  try {
    frame = JSON.parse(raw);
    if (!isObject(frame)) throw new Error("frame must be an object");
  } catch (_error) {
    process.stderr.write("invalid LAP frame\n");
    process.exitCode = 2;
    stopped = true;
    input.close();
    return;
  }
  if (!handle(frame)) {
    stopped = true;
    input.close();
  }
});
